pkg/healthmonitor/health_monitor.go

```diff
@@ -0,0 +1,338 @@
+		sm.unhealthyTargets = currentlyUnhealthyTargetsSet.List()
+		klog.V(2).Infof("observed the following unhealthy targets %v", sm.unhealthyTargets)
+		logUnhealthyTargets(sm.unhealthyTargets, currentHealthCheckProbes)
+
+		exportedUnhealthyTargets := make([]string, len(sm.unhealthyTargets))
+		for index, unhealthyTarget := range sm.unhealthyTargets {
+			exportedUnhealthyTargets[index] = unhealthyTarget
+		}
+		sm.exportedUnhealthyTargets.Store(exportedUnhealthyTargets)
+		notifyListeners = true
+	}
+
+	// detect healthy targets
+	previouslyHealthyTargetsSet := sets.NewString(sm.healthyTargets...)
+	currentlyHealthyTargetsSet := previouslyHealthyTargetsSet.Union(newHealthyTargetsSet)
+	currentlyHealthyTargetsSet.Delete(newUnhealthyTargetsSet.List()...)
+	if !currentlyHealthyTargetsSet.Equal(previouslyHealthyTargetsSet) {
+		sm.healthyTargets = currentlyHealthyTargetsSet.List()
+		klog.V(2).Infof("observed the following healthy targets %v", sm.healthyTargets)
+
+		exportedHealthyTargets := make([]string, len(sm.healthyTargets))
+		for index, healthyTarget := range sm.healthyTargets {
+			exportedHealthyTargets[index] = healthyTarget
+		}
+		sm.exportedHealthyTargets.Store(exportedHealthyTargets)
+		notifyListeners = true
+	}
+
+	if notifyListeners {
+		// notify listeners about the new healthy/unhealthy targets
+		for _, listener := range sm.listeners {
+			listener.Enqueue()
+		}
+	}
+}
+
+func (sm *HealthMonitor) healthCheckSingleTarget(target string) error {
+	// TODO: make the protocol, port and the path configurable
+	url := fmt.Sprintf("https://%s/%s", target, "readyz")
+	newReq, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := sm.client.Do(newReq)
+	if err != nil {
+		return err
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status from %v: %v, expected HTTP 200", url, resp.StatusCode)
+	}
+
+	return err
+}
+
+func createHealthCheckHTTPClient(responseTimeout time.Duration, restConfig *rest.Config) (*http.Client, error) {
+	transportConfig, err := restConfig.TransportConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig, err := transport.TLSConfigFor(transportConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: utilnet.SetTransportDefaults(&http.Transport{
+			TLSClientConfig: tlsConfig,
+		}),
+		Timeout: responseTimeout,
+	}
+
+	return client, nil
+}
+
+func logUnhealthyTargets(unhealthyTargets []string, currentHealthCheckProbes []targetErrTuple) {
+	for _, unhealthyTarget := range unhealthyTargets {
+		errorsForUnhealthyTarget := []error{}
+		for _, svrErrTuple := range currentHealthCheckProbes {
+			if svrErrTuple.target == unhealthyTarget {
+				errorsForUnhealthyTarget = append(errorsForUnhealthyTarget, svrErrTuple.err)
+			}
+		}
+		klog.V(2).Infof("the following target %v became unhealthy due to %v", unhealthyTarget, utilerrors.NewAggregate(errorsForUnhealthyTarget).Error())
```

Comment:
we probably want to turn this into a Prometheus metrics at some point

Reply:
yes, already on the TODO list https://github.com/openshift/library-go/pull/1000/commits/ed4d6682c69d3a81e4a6c2bd02144a5ba4e89d60#diff-723588c49708ae3a6e8ea0f3974fb4f2c0ba5ed298630a546a4cdb7f8e746aeeR207